neo/SmartContract/Native/Tokens/NeoToken.cs

```diff
@@ -0,0 +1,291 @@
+using Neo.Persistence;
+using Neo.VM;
+using Neo.VM.Types;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Numerics;
+using VMArray = Neo.VM.Types.Array;
+
+namespace Neo.SmartContract.Native.Tokens
+{
+    public sealed class NeoToken : Nep5Token<NeoToken.AccountState>
+    {
+        public override string ServiceName => "Neo.Native.Tokens.NEO";
+        public override string Name => "NEO";
+        public override string Symbol => "neo";
+        public override int Decimals => 0;
+        public BigInteger TotalAmount { get; }
+
+        private const byte Prefix_Validator = 33;
+        private const byte Prefix_ValidatorsCount = 15;
+
+        private NeoToken()
+        {
+            this.TotalAmount = 100000000 * Factor;
+        }
+
+        protected override StackItem Main(ApplicationEngine engine, string operation, VMArray args)
+        {
+            switch (operation)
+            {
+                case "initialize":
+                    return Initialize(engine, new UInt160(args[0].GetByteArray()));
+                case "unclaimedGas":
+                    return UnclaimedGas(engine, new UInt160(args[0].GetByteArray()), (uint)args[1].GetBigInteger());
+                case "registerValidator":
+                    return RegisterValidator(engine, args[0].GetByteArray());
+                case "vote":
+                    return Vote(engine, new UInt160(args[0].GetByteArray()), ((VMArray)args[1]).Select(p => p.GetByteArray().AsSerializable<ECPoint>()).ToArray());
+                case "getValidators":
+                    return GetValidators(engine).Select(p => (StackItem)p.ToArray()).ToArray();
+                default:
+                    return base.Main(engine, operation, args);
+            }
+        }
+
+        protected override BigInteger TotalSupply(ApplicationEngine engine)
+        {
+            return TotalAmount;
+        }
+
+        protected override void OnBalanceChanging(ApplicationEngine engine, UInt160 account, AccountState state, BigInteger amount)
+        {
+            DistributeGas(engine, account, state);
+            if (amount.IsZero) return;
+            if (state.Votes.Length == 0) return;
+            foreach (ECPoint pubkey in state.Votes)
+            {
+                StorageItem storage_validator = engine.Service.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_Validator, pubkey.ToArray()));
+                ValidatorState state_validator = ValidatorState.FromByteArray(storage_validator.Value);
+                state_validator.Votes += amount;
+                storage_validator.Value = state_validator.ToByteArray();
+            }
+            StorageItem storage_count = engine.Service.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_ValidatorsCount));
+            ValidatorsCountState state_count = ValidatorsCountState.FromByteArray(storage_count.Value);
+            state_count.Votes[state.Votes.Length - 1] += amount;
```

Comment:
Why storing on last position?

Reply:
It's `state.Votes.Length`, not `state_count.Votes.Length`.